src/macaron/repo_finder/repo_finder_base.py

```diff
@@ -0,0 +1,83 @@
+# Copyright (c) 2023 - 2023, Oracle and/or its affiliates. All rights reserved.
+# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/.
+
+"""This module contains the base class for the repo finders."""
+
+from abc import ABC, abstractmethod
+from collections.abc import Iterator
+
+
+class BaseRepoFinder(ABC):
```

Comment:
These questions are more about the interface of this class rather than the implementation.

1. I see that the `find_repo` method has the signature of

```
def find_repo(self, group: str, artifact: str, version: str)
```

which, at first glance, looks to me like it is only designed for Java. In the theme of making things language-agnostic, do you think we can change the signature of this method to make it look more language-agnostic? For example, instead of taking GAV, can the function take something like PURL as param?

2. I'm not sure if the methods `create_urls`, `retrieve_metadata`, and `read_metadata` should be part of the common interface of repo finders, for a couple of reasons:
  * In general, I don't really see why these methods should be used polymorphically outside of the repo finder packages, they seems more like internal implementation to me. They do not seem to be useful from the perspective of the consumer of the interface, e.g. if I want to find the repo of some software component, why do I need to care about creating URLs and retrieving different kinds of data? It's more likely that I just want to provide an identifier (i.e. PURL) and get a repo.
  * I also feel like these abstract methods are creating unnecessary abstractions that make the concrete implementation of each repo finder type more confusing. As an example, it's not clear to me why we should have the very board concept of `metadata` while we can have `retrieve_json_payload`/`read_json_payload` for `RepoFinderDepsDev`, and `retrieve_pom`/`read_pom` for `JavaRepoFinder` instead. These kinds of abstractions, imo, make the implementation looks kind of awkward and harder to read. For example, the `BaseRepoFinder::create_urls` abstract method returns a list of URLs, but the concrete method `RepoFinderDepsDev::create_urls` repo finder only either return 0 or 1 URL, and it is not clear why a list should the return type instead of `str | None`. In another case, it's not too clear to me what the actual return values of the `JavaRepoFinder::read_metadata` and `RepoFinderDepsDev::read_metadata` have in common, while the former is from pom contents and the latter is from request payloads.